File: src/Lorex.Api/Features/Media/ImageCrop.cs
namespace Lorex.Api.Features.Media;

/// <summary>
/// The square a thumbnail is cut from, as fractions of the picture rather than pixels of any
/// screen: <paramref name="X"/> and <paramref name="Width"/> of its width, <paramref name="Y"/> and
/// <paramref name="Height"/> of its height, from the top-left corner.
///
/// Fractions are what make the crop the picture's rather than the browser's. The same four
/// numbers select the same pixels however large the cropper was drawn, and whether the server is
/// rendering the thumbnail now or regenerating it from the stored original years later.
///
/// "The picture" means it as it is displayed: after its EXIF orientation for a JPEG or a PNG, and
/// as stored for a WebP - which is how browsers draw each one, so the square someone frames is
/// the square that gets cut. <see cref="ImagePreparation"/> owns that rule.
/// </summary>
public sealed record ImageCrop(double X, double Y, double Width, double Height);

/// <summary>
/// How long the columns that describe a stored image are allowed to be. Shared, because an
/// entry's picture and a profile photo are described by the same three strings and there is no
/// reason for two tables to disagree about their widths.
/// </summary>
public static class StoredImageLimits
{
    /// <summary>Comfortably longer than either key convention can ever produce.</summary>
    public const int ObjectKeyMaxLength = 400;

    public const int ContentTypeMaxLength = 100;

    public const int FileNameMaxLength = 255;
}

File: src/Lorex.Web/src/lib/imageCrop.ts
/**
 * The square a thumbnail is cut from, as fractions of the picture rather than pixels of any
 * screen: `x` and `width` of its width, `y` and `height` of its height, from the top-left corner.
 * The same four numbers select the same pixels however large the cropper happened to be drawn.
 *
 * Shared rather than owned by a feature, because the cropper is: an entry's portrait and an
 * account's photo are framed by the same dialog and cut by the same code on the server.
 */
export interface ImageCrop {
  x: number
  y: number
  width: number
  height: number
}

/** What may be picked in the file dialog. The API decides again by decoding the bytes. */
export const IMAGE_ACCEPT = 'image/jpeg,image/png,image/webp'

/** Mirrors the API's own ceiling, so an obviously oversized file is refused without a round trip. */
export const IMAGE_MAX_BYTES = 8 * 1024 * 1024
